File: LeetCode/19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.go
package main

import (
	"fmt"
)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func createList(l []int) *ListNode {
	if len(l) == 0 {
		return nil
	}

	head := &ListNode{Val: l[0]}
	cur := head
	for i := 1; i < len(l); i++ {
		cur.Next = &ListNode{Val: l[i]}
		cur = cur.Next
	}

	return head
}

func showList(head *ListNode) string {
	r := ""
	if head == nil {
		return r
	}

	cur := head
	for cur.Next != nil {
		r += fmt.Sprintf("%d->", cur.Val)
		cur = cur.Next
	}
	r += fmt.Sprintf("%d", cur.Val)
	return r
}

//	func showList(head *ListNode) string {
//		var builder strings.Builder
//		cur := head
//		for cur.Next != nil {
//			builder.WriteString(fmt.Sprintf("%d->", cur.Val))
//			cur = cur.Next
//		}
//		builder.WriteString(fmt.Sprintf("%d", cur.Val))
//		return builder.String()
//	}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Next: head}
	cur := dummy.Next
	total := 0
	for cur != nil {
		total++
		cur = cur.Next
	}

	prev := dummy
	cur = dummy.Next
	for i := 0; i < total-n; i++ {
		prev = cur
		cur = cur.Next
	}
	prev.Next = cur.Next

	return dummy.Next
}

func main() {
	inputs1 := [][]int{
		[]int{1, 2, 3, 4, 5},
		[]int{1},
		[]int{1, 2},
		[]int{1, 2},
	}
	inputs2 := []int{2, 1, 1, 2}
	outputs := [][]int{
		[]int{1, 2, 3, 5},
		[]int{},
		[]int{1},
		[]int{2},
	}

	for pos := 0; pos < len(outputs); pos++ {
		head := createList(inputs1[pos])
		headResult := showList(head)

		a1 := showList(removeNthFromEnd(head, inputs2[pos]))
		a2 := showList(createList(outputs[pos]))

		if a1 != a2 {
			fmt.Println(headResult, a1, a2, a1 == a2)
		}
	}

}

File: LeetCode/19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.py
from typing import List, Optional

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class ListHelper:
    def create(l: List[int]) -> Optional[ListNode]:
        if not l:
            return None

        head = ListNode(l[0])

        cur = head
        for v in l[1:]:
            cur.next = ListNode(v)
            cur = cur.next

        return head

    def traverse(head: Optional[ListNode]) -> str:
        s = ""
        if head == None:
            return s

        cur = head
        while cur.next != None:
            s += f"{cur.val}->"
            cur = cur.next
        s += f"{cur.val}"
        return s


class Solution:
    # # normal
    # def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
    #     total = 0
    #     cur = head
    #     while cur != None:
    #         total += 1
    #         cur = cur.next

    #     prev = None
    #     cur = head
    #     for _ in range(1, total+1-n):
    #         prev = cur
    #         cur = cur.next

    #     if cur == head:
    #         if cur.next is None:
    #             head = None
    #         else:
    #             head = cur.next
    #     else:
    #         prev.next = cur.next

    #     return head

    # dummy node
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head

        total = 0
        cur = head
        while cur != None:
            total += 1
            cur = cur.next

        prev = dummy
        cur = dummy.next
        for _ in range(1, total+1-n):
            prev = cur
            cur = cur.next

        prev.next = cur.next
        return dummy.next

    # two pointer
    # def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
    #     dummy = ListNode(0)
    #     dummy.next = head
    #     first = dummy
    #     second = dummy

    #     for _ in range(n+1):
    #         first = first.next

    #     while first != None:
    #         first = first.next
    #         second = second.next

    #     second.next = second.next.next

    #     return dummy.next


if __name__ == '__main__':
    sol = Solution()
    inputs1 = [[1, 2, 3, 4, 5], [1], [1, 2], [1, 2]]
    inputs2 = [2, 1, 1, 2]
    outputs = [[1, 2, 3, 5], [], [1], [2]]

    for pos in range(len(outputs)):
        head = ListHelper.create(inputs1[pos])
        h = ListHelper.traverse(head)
        ans = sol.removeNthFromEnd(head, inputs2[pos])
        a1 = ListHelper.traverse(ans)
        a2 = ListHelper.traverse(ListHelper.create(outputs[pos]))
        if a1 != a2:
            print(h, a1, a2, a1 == a2)
        # if ans != outputs[pos]:
        #     print(f'{inputs1[pos]}, {inputs2[pos]}, {ans} {outputs[pos]}')
